Remove `laminas/laminas-config` dependency

gsteel opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

With reference to [TSC issue #68](https://github.com/laminas/technical-steering-committee/issues/68), this pull removes the direct dependency on `laminas/laminas-config`

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

gsteel committed:
Replace usage of Laminas\Config with ArrayObject

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Remove direct dev dependency on laminas-config

Signed-off-by: George Steel <george@net-glue.co.uk>

weierophinney commented:
Since this changes dependencies, the patch should target next minor; I've updated it to target the 2.13.0 milestone. You'll need to rebase, I think.

weierophinney commented:
This is going to be a bit harder than replacing `Laminas\Config\Config` with `ArrayObject`. The navigation helper tests use `Laminas\Config\Factory::fromFile()` with **an XML config file** to seed the test cases. We would need to redo the assets those tests use as pure PHP, JSON, or by writing a quick XML parser. As such, I'm going to remove the milestone from this for now.

gsteel commented:
> This is going to be a bit harder than replacing `Laminas\Config\Config` with `ArrayObject`. The navigation helper tests use `Laminas\Config\Factory::fromFile()` with **an XML config file** to seed the test cases. We would need to redo the assets those tests use as pure PHP, JSON, or by writing a quick XML parser. As such, I'm going to remove the milestone from this for now.

I've just been looking at these - not sure how I didn't spot that in the first place. Is it worth considering splitting the navigation related helpers into their own package?

froschdesign commented:
@gsteel 
You can convert the navigation configuration into a PHP array. laminas-config can help here. 😄

froschdesign commented on test/Helper/Navigation/LinksTest.php:
```diff
@@ -167,7 +167,7 @@ public function testDetectRelationFromArrayPropertyOfActivePage(): void
     public function testDetectRelationFromConfigInstancePropertyOfActivePage(): void
```
The method name for this test should be updated as "Config" is no longer used.

froschdesign commented on test/Helper/Navigation/LinksTest.php:
```diff
@@ -214,7 +214,7 @@ public function testDetectMultipleRelationsFromConfigPropertyOfActivePage(): voi
         $active = $this->_helper->findOneByLabel('Page 2');
 
-        $active->addRel('alternate', new Config\Config([
+        $active->addRel('alternate', new ArrayObject([
```
The method name for this test should be updated as "Config" is no longer used.

gsteel replied:
> @gsteel
> You can convert the navigation configuration into a PHP array. laminas-config can help here. 😄

Ha ha!!! Yes - that's what I'm using now!

gsteel committed:
Replaces xml test configuration for Navigation helpers with a PHP array, cleans up imports in the abstract test and fixes a few psalm errors

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Instead of instantiating the various nav view helpers with a string, add a setup method to each test case and initialise the correct helper variant. Solves a number of psalm errors. Also use regular PHPUnit mocks in NavigationTest.php

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Annotate array type for psalm

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
- Verify that setServerUrl does require a valid url - this covers the View exception that is thrown and re-enables previously skipped test, solving a number of psalm errors.
- adds ext-dom as a dev dependency because this test uses DOMDocument

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Suppress MissingConstructor for child classes of TestCase because the PHPUnit plugin does not appear to suppress it automatically

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
- Renames test methods that no longer use Laminas\Config appropriately.
- Improves lost coverage of Variables to make sure that `toArray()` is called on objects with the expected method.
- Fixes or suppresses several psalm errors

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
CS Fix

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented:
I haven't updated the lock file or changed the target to 2.14 - it's currently even with 2.13 and the lock file is likely to cause merge conflicts if anything else gets merged. I also haven't updated psalm baseline ~200 errors have been fixed.

Ocramius reviewed (approved):
:ship: awesome, thanks @gsteel!

Ocramius commented on composer.json:
```diff
@@ -25,10 +25,10 @@
         "laminas/laminas-authentication": "^2.5",
         "laminas/laminas-cache": "^2.6.1",
         "laminas/laminas-coding-standard": "~1.0.0",
-        "laminas/laminas-config": "^3.7",
```
Oh, this change does indeed require a `composer.lock` change - I'll see if I can do it...

gsteel replied:
I can update the lock now if it suits - I have 7.3 handy

Ocramius replied:
@gsteel yes please: a `composer update laminas/laminas-config` should suffice to keep the `composer.lock` diff minimal

Ocramius replied:
Cool, `laminas/laminas-config` is done from here as a direct dependency! 

Still in there, but the `composer.lock` update is correct, so far :-)

```
root@f5e835d0a0ab:/app# composer why laminas/laminas-config
laminas/laminas-modulemanager  2.10.2  requires  laminas/laminas-config (^3.4) 
```

gsteel committed:
Update composer lock file

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius commented:
Thanks @gsteel!
